File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <unordered_set>
#include <algorithm>

using namespace std;

int solution(string begin, string target, vector<string> w) {

    unordered_set<string> s(w.begin(), w.end());
    
    if (s.find(target) == s.end()) return 0;

    queue<string> q;
    q.push(begin);
    int cnt = 0;

    while (!q.empty()) {
        int size = q.size();
        while (size--) {
            string cur = q.front();
            q.pop();

            if (cur == target) return cnt;

            for (int i = 0; i < cur.size(); i++) {
                char origin = cur[i];
                for (char c = 'a'; c <= 'z'; c++) {
                    if (c == origin) continue;
                    cur[i] = c;
                    if (s.find(cur) != s.end()) {
                        q.push(cur);
                        s.erase(cur);  
                    }
                }
                cur[i] = origin;  
            }
        }
        cnt++;
    }

    return 0; 
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        return go(words, begin, target);
    }
    
    private static int go(String[] words, String begin, String target) {
        Queue<Node> q = new LinkedList<>();
        Set<String> set = new HashSet<>();
        
        q.offer(new Node(begin, 0));
        set.add(begin);
        
        while (!q.isEmpty()) {
            Node cur = q.poll();
            
            if (cur.str.equals(target)) {
                return cur.level;
            }
            
            for (String word : words) {
                if (!set.contains(word) && check(cur.str, word)) {
                    q.offer(new Node(word, cur.level + 1));
                    set.add(word);
                }
            }
            
        }
        

        
        return 0;
    }
    
    private static boolean check(String word, String target) {
        int cnt = 0;
        for (int i = 0; i < word.length(); i++) {
           if (word.charAt(i) == target.charAt(i)) {
                cnt++;
           }
        }
            
        return cnt == word.length() - 1;
    }
    
    static class Node {
        String str;
        int level;
        
        public Node (String str, int level) {
            this.str = str;
            this.level = level;
        }
    }
}